### PR TITLE
Add minimum and maximum validations

### LINKED ISSUE #53: Review and merge all upstream's pull requests
**List of PRs**

- [x] atombender/go-jsonschema#47 
- [x] atombender/go-jsonschema#45
- [x] atombender/go-jsonschema#43
- [x] atombender/go-jsonschema#40
- [x] atombender/go-jsonschema#39
- [x] atombender/go-jsonschema#23
- [x] atombender/go-jsonschema#21

**Notes**

This is a port of my former fork's issue https://github.com/omissis/go-jsonschema-fork/issues/3

### PR DESCRIPTION
**Pull request by jmaeso**

### CODE REVIEW

**Comment by jmaeso** [comment]
Still needed to test some edge cases like float precisions and integer validations. Too late today >.<

File: pkg/schemas/model.go [hunk 1]
```diff
@@ -73,9 +73,9 @@ type Type struct {
 	Ref     string `json:"$ref,omitempty"`    // section 7
 	// RFC draft-wright-json-schema-validation-00, section 5
 	MultipleOf           int              `json:"multipleOf,omitempty"`           // section 5.1
-	Maximum              int              `json:"maximum,omitempty"`              // section 5.2
+	Maximum              float64          `json:"maximum,omitempty"`              // section 5.2
```
Comment (jmaeso) [inline pkg/schemas/model.go]: From the specification itself... http://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.6.2.2 (and same for minimum)

File: pkg/schemas/model.go [hunk 2]
```diff
@@ -73,9 +73,9 @@ type Type struct {
...
 	ExclusiveMaximum     bool             `json:"exclusiveMaximum,omitempty"`     // section 5.3
```
Comment (jmaeso) [inline pkg/schemas/model.go]: Exclusives are not following the spec but I just ignored that issue, could be fixed though :D

File: pkg/schemas/model.go (hunk 1 above)
Comment (jmaeso) [reply pkg/schemas/model.go]: btw, all section numbers are wrong

**Comment by jmaeso** [comment]
There are two different things to note at that point:
1- Blocked since I can't print to the file the character `%` for the `multipleOf` validation
2- Found a bug on my PR itself where multiple number validations can't be applied at once

File: pkg/schemas/model.go (hunk 2 above)
Comment (jmaeso) [reply pkg/schemas/model.go]: solved now

**Comment by jmaeso** [comment]
Now there is just the issue with the `%` sign and spaces on its sides.
Don't know how to solve it. 
cc/ @atombender

File: pkg/schemas/model.go (hunk 1 above)
Comment (atombender) [reply pkg/schemas/model.go]: All the section numbers refer to the quoted draft. We should sync with the latest draft.

Comment (jmaeso) [reply pkg/schemas/model.go]: I could do it, but probably in another PR. This one is focusing on the validations feature. Blocked by [this](https://github.com/atombender/go-jsonschema/pull/10#issuecomment-584063481) weird bug which I don't know how to solve/avoid

**Comment by atombender** [comment]
@jmaeso I'm interested in merging this if you could maybe rebase it from master?

**Comment by jmaeso** [comment]
I was testing after applying the rebase and got some issues. When reviewing the errors actually realized that master tests are not passing for me (suffix `-Json` is being added in all the generated structs and looks like it's not the expected behaviour). Since my code is like 1 year old now I would feel more confident if master gets fixed first in order to continue the rebase.

In case I am the only one experiencing the issue, I'm running `go version go1.15.8 darwin/amd64`

**Comment by jmaeso** [comment]
Implementation updated @atombender . Actually looks like I didn't need the `%` operator... 
TIL `%`  operator does not support floats: https://play.golang.org/p/H3yaNpKbJhu

ps: feel free to require a different implementation strategy

**Comment by scop** [comment]
Oh, incidentally I happened to hack on primitive validation too last night without noticing this, my approach is in #21 (only maximum/exclusiveMaximum done for now, but should be mostly trivial to add the rest). No personal preferences as to which approach to adopt as long as one is merged :) Will leave a few comments here.

File: pkg/generator/validator.go [hunk 3]
```diff
@@ -105,3 +107,83 @@ func (v *defaultValidator) desc() *validatorDesc {
...
+	}
+	return &validatorDesc{
+		hasError:            true,
+		beforeJSONUnmarshal: false,
+		requiredImports:     requiredImports,
+	}
+}
+
+func (v *numericValidator) generateMultipleOf(out *codegen.Emitter) {
+	out.Println(`if math.Mod(%s.%s, %f) != 0 {`, varNamePlainStruct, v.fieldName, *v.multipleOf)
```
Comment (scop) [inline pkg/generator/validator.go:152]: IIUC non-required fields are generated as pointers and this would not handle them properly (would e.g. `Mod(*float64,float64`), and no nil pointer protection). But glad to be proven wrong if that's the case! Likewise for all other `generate*` added below. I think the approach in #21 addresses this issue.

File: pkg/schemas/model.go [hunk 4]
```diff
@@ -91,11 +91,11 @@ type Type struct {
...
 	Ref     string `json:"$ref,omitempty"`    // section 7
 	// RFC draft-wright-json-schema-validation-00, section 5
-	MultipleOf           int              `json:"multipleOf,omitempty"`           // section 5.1
-	Maximum              float64          `json:"maximum,omitempty"`              // section 5.2
-	ExclusiveMaximum     bool             `json:"exclusiveMaximum,omitempty"`     // section 5.3
-	Minimum              float64          `json:"minimum,omitempty"`              // section 5.4
-	ExclusiveMinimum     bool             `json:"exclusiveMinimum,omitempty"`     // section 5.5
+	MultipleOf           *float64         `json:"multipleOf,omitempty"`           // section 5.1
+	Maximum              *float64         `json:"maximum,omitempty"`              // section 5.2
+	ExclusiveMaximum     *float64         `json:"exclusiveMaximum,omitempty"`     // section 5.3
```
Comment (scop) [inline pkg/schemas/model.go:96]: Changing Exclusive* to floats here means dropping support for draft 4 in this respect. No objections as long as it's a conscious choice. Maybe someone can figure out a clean (enough) way to support both 4 and newer? Union types would help.

**Comment by omissis** [comment]
I am closing this PR as it's become stale, and we'll likely review the whole validation system in the future.
